### ISSUE TITLE
Can't use updated "default" audio input device

### ISSUE THREAD

**Issue by vsimeikoKG** [aws/amazon-chime-sdk-js#1317]
**Meeting and attendee IDs:**
- Meeting ID: 001e8ba2-9c42-45b4-b505-64404ad4da79
- Attendee ID: 5c1d8947-9abd-bae5-be25-b4ec19002a77

**Timeline**
- User joins the meeting muted with "default" audio input selected.
- User remains muted and connects a bluetooth headset while still in the meeting.
- User can not use "default" audio input device since it was auto changed.
- User may use connected bluetooth device, but not as a "default" option though.

**Describe the bug**
Users can't use "default" audio input device if it gets changed while in the meeting.

```js
// User joins the meeting.
audioVideo.realtimeMuteLocalAudio();
await audioVideo.chooseAudioInputDevice('default');
audioVideo.start();

// User joins the meeting successfully.
// After some time new bluetooth device is connected and set as "default" option.
await audioVideo.chooseAudioInputDevice('default'); // doesn't work
await audioVideo.chooseAudioInputDevice((await audioVideo.listAudioInputDevices())[0]); // doesn't work

// User tries to say something, but he/she can't.
audioVideo.realtimeUnmuteLocalAudio();
```

`chooseAudioInputDevice` get's never resolved.

**Platform**
 - OS: macOS Catalina 10.15.7 (19H114)
 - Browser: Chrome
 - Browser Version: Version 90.0.4430.212 (Official Build) (x86_64)
 - Amazon Chime SDK GitHub Commit: 2.9.0

[localhost-1622036303591.log](https://github.com/aws/amazon-chime-sdk-js/files/6547088/localhost-1622036303591.log)

**Comment by michhyun1** [comment]
@vsimeikoKG Thanks for using the Chime SDK for Javascript. Could you verify the device that you are choosing? The bluetooth device should become the new 'default' device, but you should still be able to call 'listAudioInputDevices' and see a list of other audio input devices that are NOT the bluetooth device. 

It looks like you always choose `audioVideo.listAudioInputDevices())[0]` - maybe you can try `console.log(audioVideo.listAudioInputDevices())` to verify what devices are actually available to choose from before you call chooseAudioInputDevice. Then, you may want to get the deviceId and pass the deviceId to chooseAudioInputDevice.

**Comment by vsimeikoKG** [comment]
I can see new device when I call `listAudioInputDevices()`, but I can't choose it if I try to pass `default` or `deviceId` strings.

Audio input works perfectly if I refresh the web page and rejoin the same meeting.

![image](https://user-images.githubusercontent.com/72414337/119710202-1845d400-be2c-11eb-9d12-33622718c3dc.png)

-----

`audioVideo.listAudioInputDevices())[0]` - I wrote it as an example to indicate that I tried to pass `deviceId`  as a string and `InputDeviceInfo` object as well. Both of these options do not work without page refresh.

**Comment by michhyun1** [comment]
So if you just connected a pair of bluetooth device for audio input, I see that the default here is the bluetooth device. If you choose the default, you'll be choosing the current device that is in use. Can you show me the logs when you join a meeting with the bluetooth device as the default, and then while the meeting is running, you do:

```
const devices = await audioVideo.listAudioInputDevices();
await audioVideo.chooseAudioInputDevice(devices[1].deviceId)
```

**Comment by vsimeikoKG** [comment]
I added some comments here as well. Hope that helps!

![image](https://user-images.githubusercontent.com/72414337/119714199-94dab180-be30-11eb-8261-e884188814f2.png)

**Comment by michhyun1** [comment]
@vsimeikoKG your last comment says "Audio works now, but we want to keep using the same default audio device without forcing user to select other device"

1. So by doing this, you were able to successfully switch from the bluetooth device to the macbook pro mic, correct? If so, then the chooseAudioInputDevice is working as expected.
2. What exactly is your use case then? Are you saying that you don't want the SDK to automatically switch the default to the bluetooth device when it is connected?

**Comment by michhyun1** [comment]
If you're using the meeting demo - this is where we "reselect" the default device. In our case, when the bluetooth device is connected, it becomes the default device and we select that device as the audio input device here: https://github.com/aws/amazon-chime-sdk-js/blob/3700c741366292079833ff8bd27424dcefe2f681/demos/browser/app/meetingV2/meetingV2.ts#L1076-L1081

**Comment by vsimeikoKG** [comment]
1) `chooseAudioInputDevice` works correctly if I try to use any other device. I can not re-select same `default` device.
2) Here is the breakdown of the issue and the desired behaviour:
a) User joins the meeting with default audio input from webcam microphone. This is his current `default` audio input device.
b) If user tries to say something right now, it works correctly.
c) User connects bluetooth headset while in the meeting. This device becomes his new `default` audio input.
d) User can not use `default` device now since it has changed. User needs to refresh page OR select any other device. At this point `default` audio input device becomes "dead" option.

Desired result is that when `default` audio input gets changed, we want to continue to use the updated default device without forcing user to use another device.

**Comment by michhyun1** [comment]
ListAudioInputDevice uses the WebRTC's enumerateDevices API behind the scenes. The SDK does not decide what becomes the 'default' audio device - we are getting that from calling `await navigator.mediaDevices.enumerateDevices();` [here](https://github.com/aws/amazon-chime-sdk-js/blob/22b82e54beaa1aca6bfb583903edf22f31f86e4e/src/devicecontroller/DefaultDeviceController.ts#L1044).

If you'd like, you can do the following steps:
1. Join a meeting
2. Verify the audio input devices available by calling `await navigator.mediaDevices.enumerateDevices();`
3. Verify the default device
4. Connect Bluetooth device
5. Call `await navigator.mediaDevices.enumerateDevices();` again and verify the default device.

If you'd like to implement some logic of not selecting the default device (or any other logic around device selection on device changes), then you can feel free to comment out the section mentioned above (or provide your own logic in your application's `onAudioInputsChanged` callback method:
https://github.com/aws/amazon-chime-sdk-js/blob/3700c741366292079833ff8bd27424dcefe2f681/demos/browser/app/meetingV2/meetingV2.ts#L1076-L1081

However, the SDK doesn't determine what is chosen as the 'default' device when a new device is connected.

**Comment by vsimeikoKG** [comment]
I get an infinite loop when I call `await audioVideo.chooseAudioInputDevice('default')`. I attached logs with `INFO` logging for that as well. Hopefully that helps!

[localhost-1622127513852.log](https://github.com/aws/amazon-chime-sdk-js/files/6554938/localhost-1622127513852.log)

**Comment by michhyun1** [comment]
@vsimeikoKG make sure you're calling listAudioInputDevice before you choose an audio input device: https://github.com/aws/amazon-chime-sdk-js#device

'Use case 2. Choose audio input and audio output devices by passing the deviceId of a MediaDeviceInfo object. Note that you need to call listAudioInputDevices and listAudioOutputDevices first.'

**Comment by michhyun1** [comment]
I'm also not sure if using the string 'default' will work for other browsers, so you may want to use the deviceId instead, just in case.

**Comment by vsimeikoKG** [comment]
Here is the INFO log with the same infinite loop issue. As you asked, I listed audio input devices before choosing device:
```js
await audioVideo.listAudioInputDevices(); // default deviceId is present
await audioVideo.chooseAudioInputDevice('default');
```

[localhost-1622214296899.log](https://github.com/aws/amazon-chime-sdk-js/files/6561701/localhost-1622214296899.log)

**Comment by michhyun1** [comment]
@vsimeikoKG can you give me some reproduction steps? 

Here is what I tried:
1. Join a meeting
2. Turn on Bluetooth headphones - verify the Bluetooth headphones are connected automatically and become the 'default' audio input device
3. call `await audioVideo.listAudioInputDevices()` and verify that bluetooth headphones are the default
4. call `await audioVideo.chooseAudioInputDevice("default")
 - Verified that nothing happens, since I am re-choosing the "default" device but the default device is already in use.
 - Verified there is no infinite loop

**Comment by michhyun1** [comment]
It should compare the previous activeDevice with the new chosen device and then stop the loop here: https://github.com/aws/amazon-chime-sdk-js/blob/89145015cb820d42128c144d0b66e43cf51b4b15/src/devicecontroller/DefaultDeviceController.ts#L1359-L1361

Not sure why it is saying that the previous activeDevice is not the same as the new chosen device. The logs aren't showing that. If you could give me specific repro steps, that would be very helpful.

**Comment by michhyun1** [comment]
Another thing you might consider is just verifying the GUM call and verify that the stream you get back is the stream you expect: https://github.com/aws/amazon-chime-sdk-js/blob/89145015cb820d42128c144d0b66e43cf51b4b15/src/devicecontroller/DefaultDeviceController.ts#L1382

Perhaps enable a breakpoint here.

**Comment by vsimeikoKG** [comment]
Here is the reproduction steps:
1) Open Join Meeting Page and macOS Bluetooth devices tab.
2) Click "Join Meeting" and right away after clicking the button click on the desired device two times to connect it.
3) Wait until bluetooth device is connected, then switch back to meeting tab.
4) In console type `await audioVideo.listAudioInputDevices()` and then `await audioVideo.chooseAudioInputDevice("default")`.
5) You will get an infinite loop that will kick you out of the meeting after some time.

Notes:
- The issue is not always reproducible. Sometimes it may not happen at all, so you may need to try multiple times to recreate it.
- The bug may happen as well without executing manual step 4. This happens extremely rarely and it happened to me and another user only two or three times during the last months. Manually switching from `default` to any other device solved our problem.

**Comment by ltrung** [comment]
@vsimeikoKG Could you reproduce the issue using our [meeting demo app](https://github.com/aws/amazon-chime-sdk-js/tree/master/demos/browser)?
Also if possible, could you checkout PR #1329 and turn the Log Level to DEBUG and send us the log?
The usual cause of this is that this check for reusing existing input failed https://github.com/aws/amazon-chime-sdk-js/blob/master/src/devicecontroller/DefaultDeviceController.ts#L1313 which can be either:
- `listAudioInput` is not called beforehand and thus, the device cache inside `DefaultDeviceController` is empty.
- deviceId is not passed in correctly.

If you could debug locally, could you add a breakpoint here https://github.com/aws/amazon-chime-sdk-js/blob/master/src/devicecontroller/DefaultDeviceController.ts#L1313 and see why it could not reuse the previously selected device?

**Comment by vsimeikoKG** [comment]
@ltrung ,

I tried to recreate the issue last week with https://webclient.app.chime.aws/, but I was not able to get same broken behaviour. It happens time to time.

-----

I installed locally latest `master` branch which has your PR #1329 merged in. I attached logs with debug level as well where the issue happens.

Note: we can detect local audio input volume through `MediaStream`, so device is working.

[master-branch-logs-debug-lvl.txt](https://github.com/aws/amazon-chime-sdk-js/files/6600397/master-branch-logs-debug-lvl.txt)

-----

I placed breakpoint on the `matchesDeviceSelection()` function declaration and tried a few times - was not able to recreate issue. The promise from `await audioVideo.chooseAudioInputDevice('default')` gets instantly resolved after I clear up breakpoints and it returns `undefined` result.

However, as soon as I remove breakpoint I can get the same bug where `await audioVideo.chooseAudioInputDevice('default')` gets never resolved.

**Comment by michhyun1** [comment]
@vsimeikoKG Thanks for the logs. I was looking at the section where the infinite loop exists, and something caught my eye:

You call chooseAudioInputDevice with the parameter 'default'. Looking at the listAudioInputDevices output, you'd expect the 'default' device to be the Bluetooth headphones.
`
listAudioInputDevices false -> [{"deviceId":"default","kind":"audioinput","label":"Default - HD 4.40BT (Bluetooth)","groupId":"d4744d904b2314df904a4275c164b0bf3298a6cdc361a7e29548c77995ac0bb8"},{"deviceId":"c09800c39fb86f01199142315450a91ff0f29831841530892b8fef367fb44654","kind":"audioinput","label":"HD 4.40BT (Bluetooth)","groupId":"d4744d904b2314df904a4275c164b0bf3298a6cdc361a7e29548c77995ac0bb8"},{"deviceId":"da9be0dddc298c4babd16561353f6bde3d640444c30b3d75fb2791792eb5b325","kind":"audioinput","label":"Capture Inactive (1d6b:0102)","groupId":"aece0939371a37a3ad36fd7e9fbc58feee190d098a7084e19704f4f1dc66ca7a"},{"deviceId":"ecaf83659eaf972d6c0942949365ae6e629765aa05bb1f7e9564ce4c26fa4fb0","kind":"audioinput","label":"MacBook Pro Microphone (Built-in)","groupId":"28894f9ec765ff5e2bb866fff06d6e67732d981e7444da0d943ace874e94e4f9"}]
`

However, it seems like what's actually happening is that when you call chooseAudioInputDevice('default') we are selecting a different device (this is in your logs): 
`
Set activeDevice to {"groupId":"aece0939371a37a3ad36fd7e9fbc58feee190d098a7084e19704f4f1dc66ca7a","stream":{},"constraints":{"audio":{"echoCancellation":true,"googEchoCancellation":true,"googEchoCancellation2":true,"googAutoGainControl":true,"googAutoGainControl2":true,"googNoiseSuppression":true,"googNoiseSuppression2":true,"googHighpassFilter":true,"deviceId":{"exact":"default"}}}}
`

This device groupId corresponds to the 'Capture Inactive' device, not the 'default' device (Default - HD 4.40BT (Bluetooth)) specified by listAudioInputDevice output. 

It seems as if getUserMedia is returning the 'Capture Inactive' device when passed in the constraints -> `deviceId === 'default'.` (This seems wrong - if your bluetooth headphones are connected and are the 'default' device according to listAudioInputDevices, then getUserMedia should return the Bluetooth headphones when asked for the 'default' device).

In the meantime, can you specify the deviceId explicitly in chooseAudioInputDevice instead of relying on the getUserMedia call to determine what the 'default' device is?

**Comment by vsimeikoKG** [comment]
@michhyun1 ,

We rewrote some logic and we implemented Singleton for MediaStreams because of iOS Safari bug.

Here is an updated hook with the same infinite loop issue:
```js
  useEffect(() => {
    if (state.meetingStatus !== MeetingStatusEnum.Loading) {
      const {audioVideo} = meetingSession.current;
     
      const handler = (stream) => {
        if (!stream && !audioVideo.realtimeIsLocalAudioMuted()) {
          audioVideo.realtimeMuteLocalAudio();
        }

        if (stream && audioVideo.realtimeIsLocalAudioMuted()) {
          const clone = stream.clone();

          console.log('CLONE', clone);

          audioVideo.chooseAudioInputDevice(clone).then(() => {
            console.log('CHOOSE DEVICE FROM STREAM HERE');
            audioVideo.realtimeUnmuteLocalAudio();
            console.log('UNMUTE HERE');
          });
        }
      };

      MediaStreamSingleton.subscribeToAudioStream(handler);

      return () => MediaStreamSingleton.unsubscribeFromVideoStream(handler);
    }
  }, [state.meetingStatus]);
```

Hopefully, it makes it a bit easier to recreate the issue with `MediaStream.clone()`.

If it doesn't help, then I will proceed with your suggested debugging steps with our old code.

[logs.txt](https://github.com/aws/amazon-chime-sdk-js/files/6626939/logs.txt)

**Comment by michhyun1** [comment]
@vsimeikoKG Which iOS Safari bug are you referring to?

Also,  I have been trying to reproduce this issue with your reproduction steps, but I haven't been able to reproduce it. In this case above, you are using a MediaStream as an input instead of the string 'default'. I also havent been able to reproduce the issue using a MediaStream as the parameter. 

Could you try one thing - we recently added a parameter for listAudioInputDevices called forceUpdate. Perhaps you can try setting the boolean to true so that it will always refresh the device cache that we maintain in the SDK. Please try that out and let us know if there are any issues. 

https://github.com/aws/amazon-chime-sdk-js/blob/5a68c57387b79c5c9361270d8a0878f6ede1ad2b/src/devicecontroller/DeviceController.ts#L57

**Comment by vsimeikoKG** [comment]
@michhyun1 ,

We faced an iOS bug where `getUserMedia()` call was erasing all previous MediaStream objects, so we implemented a simple Singleton for handling video & audio MediaStream objects. https://bugs.webkit.org/show_bug.cgi?id=179363

I tried using forceUpdate, but it didn't help:
```js
  useEffect(() => {
    if (state.meetingStatus !== MeetingStatusEnum.Loading) {
      const {audioVideo} = meetingSession.current;

      const handler = (stream) => {
        console.log('MEDIA STREAM', stream);

        if (!stream && !audioVideo.realtimeIsLocalAudioMuted()) {
          console.log('MUTE AUDIO');
          audioVideo.realtimeMuteLocalAudio();
        }

        if (stream && audioVideo.realtimeIsLocalAudioMuted()) {
          console.log('ENABLE AUDIO');

          audioVideo
            .listAudioInputDevices(true)
            .then(() =>
              audioVideo
                .chooseAudioInputDevice(stream.clone())
                .then(() => audioVideo.realtimeUnmuteLocalAudio())
            );
        }
      };

      MediaStreamSingleton.subscribeToAudioStream(handler);

      return () => MediaStreamSingleton.unsubscribeFromAudioStream(handler);
    }
  }, [state.meetingStatus]);
```

Also, I've updated chime version to 2.11.0, so it should have extra debug logs.

Here is a debug level logs as well:
[logs.txt](https://github.com/aws/amazon-chime-sdk-js/files/6651110/logs.txt)

**Comment by michhyun1** [comment]
@vsimeikoKG is there any way you can use a deviceId instead of 'default' or a MediaStream when calling chooseAudioInputDevice? It seems as if there is a bug when using a mediaStream with chooseAudioInputDevice. 

Is there a way you can implement this method using a deviceId instead?

**Comment by vsimeikoKG** [comment]
@michhyun1 ,

If I use `deviceId`, then it works correctly:
```js
audioVideo
  .chooseAudioInputDevice(stream.getAudioTracks()[0].getSettings().deviceId)
  .then(() => audioVideo.realtimeUnmuteLocalAudio());
```

Is there any plans to to get `chooseAudioInputDevice` to work with `MediaStream` objects?


P.S. There is a [similar issue](https://github.com/aws/amazon-chime-sdk-js/issues/1357) for `chooseVideoInputDevice()` when using `MediaStream`.

**Comment by michhyun1** [comment]
@vsimeikoKG yes, we recently found this bug, and we've noted it. Are you unblocked now? The same issue should exist in chooseVideoInputDevice, as they re-use similar code. Using deviceId is the suggested method for calling chooseAudio/VideoInput device, and we'll work on fixing the bug related to using MediaStream as the parameter.

**Comment by vsimeikoKG** [comment]
@michhyun1 ,

We're currently blocked by this as we need to use MediaStream outside of the Chime meeting for:
1) Video preview
2) Audio indicator
3) Audio testing

We are currently using `deviceId`, but we need to switch to the new MediaStream singleton logic that we implemented in order to fix the iOS issue.

**Comment by michhyun1** [comment]
@vsimeikoKG thanks for letting us know. We're working on fixing this bug with using the MediaStream in chooseAudioInputDevice API. If you could perhaps give us more details on why you need the singleton and what your use case is? It seems like you are calling the chooseAudioInputDevice multiple times, which calls getUserMedia behind the scenes and thus triggers the iOS bug. Have you tried stopping the stream before choosing a new stream?

**Comment by michhyun1** [comment]
@vsimeikoKG we've merged a fix for using the mediaStream as the input for chooseAudioInputDevice. It's currently sitting on the HEAD of master branch. You may try using that and see if this unblocks you.

**Comment by michhyun1** [comment]
@vsimeikoKG thats great to hear. Is there any additional issues you are facing now? If not, please feel free to close this issue out. thanks

**Comment by vsimeikoKG** [comment]
@michhyun1 ,
No issues. Closing the ticket.

**Comment by sj94123** [comment]
Facing a similar issue with Chime Client. When a user switches a default device using system settings, the getUserMedia considers the previously connected device as the default device and connects to that instead of the most recently changed default device.

**Comment by michhyun1** [comment]
@sj94123 can you please create a new issue and link to this issue if its related? Thanks
